cartographer/mapping_3d/pose_graph/optimization_problem.cc

```diff
@@ -98,6 +100,45 @@ std::unique_ptr<transform::Rigid3d> Interpolate(
   return nullptr;
 }
 
+void AddLandmarkCostFunctions(
+    const std::map<std::string, LandmarkNode>& landmark_nodes,
+    const mapping::MapById<mapping::NodeId, NodeData>& node_data,
+    mapping::MapById<mapping::NodeId, CeresPose>* C_nodes,
+    std::map<std::string, CeresPose>* C_landmarks, ceres::Problem* problem) {
+  for (const auto& landmark_node : landmark_nodes) {
+    for (const auto& observation : landmark_node.second.landmark_observations) {
+      const std::string& landmark_id = landmark_node.first;
+      // Find the trajectory nodes before and after the landmark observation.
+      auto prev =
+          node_data.lower_bound(observation.trajectory_id, observation.time);
+      if (prev == node_data.BeginOfTrajectory(observation.trajectory_id)) {
```

Comment:
BeginOfTrajectory() is a valid node. Are you sure you should skip in all cases? What if observation.time is between first and second trajectory node?